minoc/src/main/java/org/opencadc/minoc/PutAction.java

```diff
@@ -178,47 +184,60 @@ public void doAction() throws Exception {
 
         profiler.checkpoint("content.init");
 
+        // commit transaction or write data
         StorageMetadata artifactMetadata = null;
-        
-        log.debug("writing new artifact to " + storageAdapter.getClass().getName());
-        try {
-            artifactMetadata = storageAdapter.put(newArtifact, in);
-            profiler.checkpoint("storageAdapter.put.ok");
-        } catch (ReadException ex) {
-            profiler.checkpoint("storageAdapter.put.fail");
-            if (contentLength != null) {
-                Throwable cause = ex.getCause();
-                while (cause != null) {
-                    if (cause instanceof EOFException) {
-                        throw new PreconditionFailedException("premature end-of-stream: expected content-length " + contentLength, cause);
+        if (txnID != null && contentLength != null && contentLength == 0) {
+            artifactMetadata = storageAdapter.commitTransaction(txnID);
+            txnID = null;
+            profiler.checkpoint("storageAdapter.put.commit.ok");
+        } else {
+            log.debug("writing new artifact to " + storageAdapter.getClass().getName());
+            try {
```

Comment:
Just wondering if we should attempt to detect changes in Content-Length, Digest if the client tries to resume with a modified file.

Reply:
This is just single put with separate commit, not resume... so I haven't looked at detecting discrepancies between separate calls. And I can tell that resume will be more complex to implement than the design might have implied (wr.t. Digest in particular).